go/ql/lib/semmle/go/frameworks/Fasthttp.qll

```diff
@@ -0,0 +1,372 @@
+      RequestForgerySinkDial() {
+        exists(Method m |
+          m.hasQualifiedName(packagePath(), "TCPDialer",
+            ["Dial", "DialTimeout", "DialDualStack", "DialDualStackTimeout"]) and
+          this = m.getACall().getArgument(0)
+        )
+      }
+
+      override DataFlow::Node getARequest() { result = this }
+
+      override string getKind() { result = "TCP Addr + Port" }
+    }
+  }
+
+  /**
+   * Provide modeling for fasthttp.Client Type
+   */
+  module Client {
+    /**
+     * A method that sends HTTP requests.
+     * Get* send a HTTP GET request.
+     * Post send a HTTP POST request.
+     * these Functions first arguments is a URL.
+     */
+    class RequestForgerySink extends RequestForgery::Sink {
+      RequestForgerySink() {
+        exists(Method m |
+          m.hasQualifiedName(packagePath(), "Client", ["Get", "GetDeadline", "GetTimeout", "Post"]) and
+          this = m.getACall().getArgument(1)
+        )
+      }
+
+      override DataFlow::Node getARequest() { result = this }
+
+      override string getKind() { result = "URL" }
+    }
+  }
+
+  /**
+   * Provide modeling for fasthttp.HostClient Type
+   */
+  module HostClient {
+    /**
+     * A method that sends HTTP requests.
+     * Get* send a HTTP GET request.
+     * Post send a HTTP POST request.
+     * these Functions first arguments is a URL.
+     */
+    class RequestForgerySink extends RequestForgery::Sink {
+      RequestForgerySink() {
+        exists(Method m |
+          m.hasQualifiedName(packagePath(), "HostClient",
+            ["Get", "GetDeadline", "GetTimeout", "Post"]) and
+          this = m.getACall().getArgument(1)
+        )
+      }
+
+      override DataFlow::Node getARequest() { result = this }
+
+      override string getKind() { result = "URL" }
+    }
+  }
+
+  /**
+   * Provide modeling for fasthttp.Response Type
+   */
+  module Response {
+    /**
+     * A Method That send files from its input and it does not check input path against path traversal attacks, so it is a dangerous method
+     */
+    class FileSystemAccess extends FileSystemAccess::Range, DataFlow::CallNode {
+      FileSystemAccess() {
+        exists(Method mcn |
+          mcn.hasQualifiedName(packagePath(), "Response", "SendFile") and
+          this = mcn.getACall()
+        )
+      }
+
+      override DataFlow::Node getAPathArgument() { result = this.getArgument(0) }
+    }
+
+    /**
+     * The methods that can write to HTTP Response Body.
+     * These methods can be dangerous if they are user controllable.
+     */
+    class HttpResponseBodySink extends SharedXss::Sink {
+      HttpResponseBodySink() {
+        exists(Method m |
+          m.hasQualifiedName(packagePath(), "Response",
+            [
+              "AppendBody", "AppendBodyString", "SetBody", "SetBodyString", "SetBodyRaw",
+              "SetBodyStream"
+            ]) and
+          this = m.getACall().getArgument(0)
+        )
+      }
+    }
```

Comment:
```suggestion
        or
        exists(Method write, DataFlow::CallNode writeCall |
          write.hasQualifiedName("io", "Writer", "Write") and
          writeCall = write.getACall() and
          ResponseBodyWriterFlow::flowsTo(writeCall.getReceiver()) and
          this = writeCall.getArgument(0)
        )
      }
    }

    private predicate responseBodyWriterResult(DataFlow::Node src) {
      exists(Method responseBodyWriter |
        responseBodyWriter.hasQualifiedName(packagePath(), "Response", "BodyWriter") and
        src = responseBodyWriter.getACall().getResult(0)
      )
    }

    private module ResponseBodyWriterFlow = DataFlow::SimpleGlobal<responseBodyWriterResult/1>;
```
This makes the test pass. It uses `DataFlow::SimpleGlobal` which is a very new, experimental module which does interprocedural flow, but doesn't do as much work as full global flow. I think we need interprocedural flow because when I searched GitHub I found lots of examples of the return value of `Response.BodyWriter()` being passed into another function.